### PR TITLE
[VL] Updated documentation, refactoring and added more testcases for BNLJ

### PR DESCRIPTION
**Pull request by Surbhi-Vijay**
## What changes were proposed in this pull request?
1. Updated documentation for CartesianProduct and BroadcastNestedLoopJoin
2. Added fallback test cases for BNLJ
3. Renamed methods in BroadcastApi to generic names since it is being used for both HashedBroadcastMode as well as IdentityBroadcastMode (discussed [here](https://github.com/oap-project/gluten/pull/4565#discussion_r1495556405))

## How was this patch tested?
Tested using unit tests

### CODE REVIEW

**Comment by Surbhi-Vijay** [comment]
CC: @zhouyuan @zhztheplayer

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/gluten/GlutenFallbackSuite.scala [hunk 1]
```diff
@@ -93,6 +93,18 @@ class GlutenFallbackSuite extends GlutenSQLTestsTrait with AdaptiveSparkPlanHelp
...
+    withTable("t1", "t2") {
+      spark.range(10).write.format("parquet").saveAsTable("t1")
+      spark.range(10).write.format("parquet").saveAsTable("t2")
+
+      val id = runExecution("SELECT * FROM t1 FULL OUTER JOIN t2")
+      val execution = glutenStore.execution(id)
+      assert(execution.get.numFallbackNodes == 1)
+      assert(
+        execution.get.fallbackNodeToReason.head._2
+          .contains("FullOuter join is not supported with Broadcast Nested Loop Join"))
```
Comment (zhztheplayer) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/gluten/GlutenFallbackSuite.scala:105]: Nit: Probably we should use consistent casing between "full-outer join" and "broadcast nested-loop join"?

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala [hunk 2]
```diff
@@ -898,7 +898,7 @@ class TestOperator extends VeloxWholeStageTransformerSuite with AdaptiveSparkPla
       withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "1MB") {
         runQueryAndCompare(
           """
-            |select * from t1 cross join t2;
+            |select * from t1 left outer join t2;
             |""".stripMargin
         ) {
           checkOperatorMatch[BroadcastNestedLoopJoinExecTransformer]
```
Comment (zhztheplayer) [inline backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:904]: Would you like to explain a bit about the reason changing the test case? Thanks.

**Review by zhztheplayer — COMMENTED** [review]
Thanks! Just some minors.

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala (hunk 2 above)
Comment (Surbhi-Vijay) [reply backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:910]: Added expression matching to enhance the test case, previously it was not covered.

**Review by zhztheplayer — APPROVED** [review]
Thanks!

**Comment by Surbhi-Vijay** [comment]
@zhztheplayer Thank you for the review!
can you please tell, how can I rerun failed run for "centos7"?

**Comment by zhouyuan** [comment]
@Surbhi-Vijay many thanks for your big help on this feature!

**Pull request merged**